### PR TITLE
GoogleMaps reduce # requests. getXY . Fixes #1596

### LINKED ISSUE #1596: too many server requests for googleMaps

### PR DESCRIPTION
**Pull request by hamidzr**
cuts the number of calls in most google map projects to half

### CODE REVIEW

**Comment by hamidzr** [comment]
another way of handling the issue would be to change the interface that it accpets an array. 
then the user could pass latLon for one point or an array of lat longs 
`[[lat,lon],[lat,lon]]`

**Comment by brollb** [comment]
I like the idea about optionally accepting an array. I would remove `getEarthCoordinate` bc "coordinate" should be plural in both cases.

It looks like the tests don't match the added methods... that is an issue.

I would rename `getScreenCoordinates` to `getImageCoordinates` because we are actually converting to the position relative to the returned image

**Comment by hamidzr** [comment]
tests will be updated as soon as we decide on the method names.

**Comment by brollb** [comment]
I like having two methods `getImageCoordinates` and `getEarthCoordinates` which both accept two arguments and return a 2D list with the given coordinates.

Batching requirements can be done probably by accepting lists of x/y values and re-using the same method. I am really not a fan of making a special method for batching the requests bc that isn't scalable wrt all the methods (it would result in duplicating all the methods). I am a fan of leveraging a new abstraction to compose these types of batching by creating anonymous functions to run on the server #1152.

That being said, I know that part of the motivating problem here was the eclipse visualization demo as it would speed up initialization time. I wonder if it would be better to change this PR to only address the singular case (as this will cut the network requests in half) and see if that improves the initialization time enough for now (until we get the other feature implemented).

**Comment by hamidzr** [comment]
that would be reverting back to commit one for this branch. Then I would have them named as getImageCoordinates and getEarthCoordinates

File: src/server/rpc/procedures/google-maps/google-maps.js [hunk 1]
```diff
@@ -206,9 +206,39 @@ StaticMap.prototype.getLatitude = function(y){
     });
 };
 
+/**
+ * converts image coordinates to earth coordinates.
+ * @param {Number} x x position of the point on the stage
+ * @param {Number} y y position of the point on the stage
+ * @returns {Array} latitude longitude of the point as a list.
```
Comment (brollb) [inline src/server/rpc/procedures/google-maps/google-maps.js]: I would rephrase this to say "A list containing the latitude and longitude of the given point"

File: src/server/rpc/procedures/google-maps/google-maps.js [hunk 2]
```diff
@@ -206,9 +206,39 @@ StaticMap.prototype.getLatitude = function(y){
     });
 };
 
+/**
+ * converts image coordinates to earth coordinates.
+ * @param {Number} x x position of the point on the stage
```
Comment (brollb) [inline src/server/rpc/procedures/google-maps/google-maps.js]: We can't say "on the stage" because it is technically on the map image. The coordinates are actually relative to the image and not necessarily the stage.

File: src/server/rpc/procedures/google-maps/google-maps.js [hunk 3]
```diff
@@ -206,9 +206,39 @@ StaticMap.prototype.getLatitude = function(y){
...
+
+StaticMap.prototype.getEarthCoordinates = function(x, y){
+    return this._getMapInfo(this.socket.roleId).then(mapInfo => {
+        let coords = this._coordsAt(x,y, mapInfo);
+        return [coords.lat, coords.lon];
+    });
+};
+
+/**
+ * converts earth coordinates to image coordinates.
```
Comment (brollb) [inline src/server/rpc/procedures/google-maps/google-maps.js]: This should be reworded to something like "Get the image coordinates (x, y) of a given location on the earth (latitude, longitude)"

File: src/server/rpc/procedures/google-maps/google-maps.js [hunk 4]
```diff
@@ -206,9 +206,39 @@ StaticMap.prototype.getLatitude = function(y){
     });
 };
 
+/**
+ * converts image coordinates to earth coordinates.
```
Comment (brollb) [inline src/server/rpc/procedures/google-maps/google-maps.js]: Rephrase "Get the given earth coordinates (latitude, longitude) of a given point in the last requested map image (x, y)"

**Comment by hamidzr** [comment]
do we need to include details about the input in the description? isn't that what @param description is for?

File: src/server/rpc/procedures/google-maps/google-maps.js [hunk 5]
```diff
@@ -206,9 +206,39 @@ StaticMap.prototype.getLatitude = function(y){
...
+StaticMap.prototype.getEarthCoordinates = function(x, y){
+    return this._getMapInfo(this.socket.roleId).then(mapInfo => {
+        let coords = this._coordsAt(x,y, mapInfo);
+        return [coords.lat, coords.lon];
+    });
+};
+
+/**
+ * Get the image coordinates (x, y) of a given location on the earth (latitude, longitude).
+ * @param {Latitude} latitude latitude of the point on the stage
```
Comment (brollb) [inline src/server/rpc/procedures/google-maps/google-maps.js]: Remove `on the stage` as it might not be on the stage (same with the next line)

File: src/server/rpc/procedures/google-maps/google-maps.js [hunk 6]
```diff
@@ -206,9 +206,39 @@ StaticMap.prototype.getLatitude = function(y){
     });
 };
 
+/**
+ * Get the given earth coordinates (latitude, longitude) of a given point in the last requested map image (x, y).
```
Comment (brollb) [inline src/server/rpc/procedures/google-maps/google-maps.js]: I would remove the first `given` (sounds a little funny)
